src/interpreter.c

```diff
@@ -665,39 +795,56 @@ jl_value_t *jl_interpret_call(jl_method_instance_t *lam, jl_value_t **args, uint
     JL_GC_PUSHARGS(locals, jl_source_nslots(src) + jl_source_nssavalues(src) + 2);
     locals[0] = (jl_value_t*)src;
     locals[1] = (jl_value_t*)stmts;
-    interpreter_state s = {};
-    s.src = src;
-    s.module = lam->def.method->module;
-    s.locals = locals + 2;
-    s.sparam_vals = lam->sparam_vals;
+    s->src = src;
+    s->module = args->lam->def.method->module;
+    s->locals = locals + 2;
+    s->sparam_vals = args->lam->sparam_vals;
     size_t i;
-    for (i = 0; i < lam->def.method->nargs; i++) {
-        if (lam->def.method->isva && i == lam->def.method->nargs - 1)
-            s.locals[i] = jl_f_tuple(NULL, &args[i], nargs - i);
+    for (i = 0; i < args->lam->def.method->nargs; i++) {
+        if (args->lam->def.method->isva && i == args->lam->def.method->nargs - 1)
+            s->locals[i] = jl_f_tuple(NULL, &args->args[i], args->nargs - i);
         else
-            s.locals[i] = args[i];
+            s->locals[i] = args->args[i];
     }
-    jl_value_t *r = eval_body(stmts, &s, 0, 0);
+    jl_value_t *r = eval_body(stmts, s, 0, 0);
     JL_GC_POP();
-    return r;
+    return (void*)r;
 }
 
-jl_value_t *jl_interpret_toplevel_thunk(jl_module_t *m, jl_code_info_t *src)
+SECT_INTERP jl_value_t *jl_interpret_call(jl_method_instance_t *lam, jl_value_t **args, uint32_t nargs)
 {
-    jl_array_t *stmts = src->code;
+    if (lam->jlcall_api == 2)
+        return lam->inferred_const;
+    struct jl_interpret_call_args callback_args = {.lam = lam, .args = args, .nargs = nargs};
+    return (jl_value_t*)enter_interpreter_frame(jl_interpret_call_callback, (void *)&callback_args);
+}
+
+struct jl_interpret_toplevel_thunk_args {
+    jl_module_t *m;
+    jl_code_info_t *src;
+};
+SECT_INTERP void *jl_interpret_toplevel_thunk_callback(interpreter_state *s, void *vargs) {
+    struct jl_interpret_toplevel_thunk_args *args =
+        (struct jl_interpret_toplevel_thunk_args*)vargs;
+    jl_array_t *stmts = args->src->code;
     assert(jl_typeis(stmts, jl_array_any_type));
     jl_value_t **locals;
-    JL_GC_PUSHARGS(locals, jl_source_nslots(src) + jl_source_nssavalues(src));
-    interpreter_state s = {};
-    s.src = src;
-    s.locals = locals;
-    s.module = m;
-    s.sparam_vals = jl_emptysvec;
+    JL_GC_PUSHARGS(locals, jl_source_nslots(args->src) + jl_source_nssavalues(args->src));
+    s->src = args->src;
+    s->locals = locals;
+    s->module = args->m;
+    s->sparam_vals = jl_emptysvec;
     size_t last_age = jl_get_ptls_states()->world_age;
-    jl_value_t *r = eval_body(stmts, &s, 0, 1);
+    jl_value_t *r = eval_body(stmts, s, 0, 1);
     jl_get_ptls_states()->world_age = last_age;
     JL_GC_POP();
-    return r;
+    return (void*)r;
+}
+
+SECT_INTERP jl_value_t *jl_interpret_toplevel_thunk(jl_module_t *m, jl_code_info_t *src)
+{
+    struct jl_interpret_toplevel_thunk_args args = {.m = m, .src = src};
```

Comment:
This C99 syntax is not present in C++

Reply:
Are we still trying to write code that can be compiled by MSVC?

Reply:
Yes. Also supports the JULIA_TIMING option.